# sumario <a id='sumario'></a>
- [Objetivo](#objetivo)
- [Livros depois de janeiro de 2000](#2000)
- [avaliaçõese classificação media](#media)
- [editora do maior numero de livros com mais de 50 paginas](#maior)
- [autor com media mais alta entre mais de 50 classificaçoes](#autor)
- [media de avaliação dos usuarios com mais de 50 classificaçoes](#classificacao)

## Objetivo: 

-entender as tendencias de mercado para propor um novo produto

## Tarefas:

Encontre o número de livros lançados depois de 1 de janeiro de 2000.
Encontre o número de avaliações e a classificação média para cada livro.

Identifique a editora que lançou o maior número de livros com mais de 50 páginas (isso vai ajudar você a excluir brochuras e publicações parecidas da sua análise).
Identifique o autor com a média mais alta classificação de livros: olhe apenas para livros com pelo menos 50 classificações.
Encontre o número médio de avaliações entre usuários que classificaram mais do que 50 livros.
Instruções para completar a tarefa
Descreva os objetivos do estudo.
Estude as tabelas (imprima as primeiras linhas).
Faça uma consulta SQL para cada uma das tarefas.
Enuncie os resultados de cada consulta no Caderno.
Descreva suas conclusões para cada uma das tarefas.
Acessando o conjunto de dados
Conecte-se ao banco de dados seguindo essas instruções.

Anotações
Não se esqueça das funções! Eles podem fazer a sua vida e a execução de consultas bem mais fácil.
Seus resultados devem ser obtidos com SQL. Use pandas apenas para imprimir e armazene o resultado das consultas.

In [1]:
# importar bibliotecas
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'practicum_student', # nome de usuário
'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # senha
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # porta de conexão
'db': 'data-analyst-final-project-db'} # o nome do banco de dados

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})




In [2]:
books = "SELECT * FROM books;"

pd.io.sql.read_sql(books, con = engine)


book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [3]:

authors = "SELECT * FROM authors ;"

pd.io.sql.read_sql(authors, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [4]:
  
publishers = "SELECT * FROM publishers ;"

pd.io.sql.read_sql(publishers, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [5]:
ratings = "SELECT * FROM ratings ;"

pd.io.sql.read_sql(ratings, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [6]:
reviews = "SELECT * FROM reviews ;"

pd.io.sql.read_sql(reviews, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [7]:
def consulta(x):
    return pd.io.sql.read_sql(x, con = engine)
    

In [8]:
consulta(reviews)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

temos 1000 livros de 635 autores publicados por 340 editoras com 6456 avaliações e 2793 criticas

## Numero de Livros lançados depois de janeiro de 2000 <a id='2000'> </a>
[sumario](#sumario)

para consultar precisamos da tabela books

In [9]:
consulta(books)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

consulta_janeiro_2000 = "SELECT publication_date, title as livro FROM books WHERE publication_date::date >= '2000-01-01'"
consulta(consulta_janeiro_2000)

In [10]:
consulta_janeiro_2000 = "SELECT count(*) FROM books WHERE publication_date::date >= '2000-01-01'"
consulta(consulta_janeiro_2000)

count
0    821

temos 821 publicações após o ano 2000

## avaliaçõese classificação media <a id='media'></a>

Encontre o número de avaliações e a classificação média para cada livro.
para tanto usaremos duas tabelas books e rating


In [11]:
consulta(books)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [12]:
consulta(ratings)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [13]:
# 1 o número de avaliações e 2 a classificação média para cada livro.
classificacao = """
SELECT books.book_id,count(ratings.rating) as numero_de_classificacao, ROUND(AVG(ratings.rating), 1) as classificacao_media, title as livro 
FROM books 
JOIN ratings 
ON books.book_id=ratings.book_id 
GROUP BY books.book_id
ORDER BY numero_de_classificacao DESC, classificacao_media DESC """
consulta(classificacao)

book_id  numero_de_classificacao  classificacao_media  \
0        948                      160                  3.7   
1        750                       88                  4.1   
2        673                       86                  3.8   
3         75                       84                  3.7   
4        302                       82                  4.4   
..       ...                      ...                  ...   
995      939                        2                  2.5   
996      371                        2                  2.0   
997      316                        2                  2.0   
998      303                        2                  1.5   
999      191                        1                  4.0   

                                                 livro  
0                              Twilight (Twilight  #1)  
1                  The Hobbit  or There and Back Again  
2                               The Catcher in the Rye  
3                 Angels & Demons (Robert Langdon  #1)  
4    Harry Potter and the Prisoner of Azkaban (Harr...  
..                                                 ...  
995                                   Tropic of Cancer  
996                                              Junky  
997                  His Excellency: George Washington  
998                               Harvesting the Heart  
999  Disney's Beauty and the Beast (A Little Golden...  

[1000 rows x 4 columns]

In [14]:
avaliacao= f''' 
SELECT count(r.text), AVG(classificacao.classificacao_media)
from reviews r 
join({classificacao}) as classificacao 
on classificacao.book_id=r.book_id'''
consulta(avaliacao)

count       avg
0   2793  3.906158

ok encontramos os numeros de avaliações 2793 e a media das medias de classificações por livro, sendo 3,9 . Além disso, é 6,46 a quantidade de avaliações por livro. as notas vão de 0 a 5, sendo que os unicos que contem 5 estrelas têm menos de 5 avaliações, já os 3 que foram mais classificados têm 160, depois 88 e 86 com classificações proximas a 4

## editora do maior numero de livros com mais de 50 paginas<a id='maior'></a>

Identifique a editora que lançou o maior número de livros com mais de 50 páginas (isso vai ajudar você a excluir brochuras e publicações parecidas da sua análise).
para tanto será necessário trabalhar com books e com publishers

In [15]:
consulta(books)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [16]:
consulta(publishers)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [17]:
editora_livros_50pag='select publisher, count(book_id) from publishers p join books b on p.publisher_id=b.publisher_id where num_pages > 50 group by publisher order by count desc'
editoratop10='select publisher, count(book_id) from publishers p join books b on p.publisher_id=b.publisher_id where num_pages > 50 group by publisher order by count desc limit 10'

consulta(editoratop10)

publisher  count
0              Penguin Books     42
1                    Vintage     31
2   Grand Central Publishing     25
3           Penguin Classics     24
4           Ballantine Books     19
5                     Bantam     19
6                    Berkley     17
7         St. Martin's Press     14
8              Berkley Books     14
9  William Morrow Paperbacks     13

In [18]:
consulta(editora_livros_50pag)

publisher  count
0                                        Penguin Books     42
1                                              Vintage     31
2                             Grand Central Publishing     25
3                                     Penguin Classics     24
4                                     Ballantine Books     19
..                                                 ...    ...
329                                         Turtleback      1
330  Atheneum Books for Young Readers: Richard Jack...      1
331                                     Penguin Signet      1
332                                    Victor Gollancz      1
333                      Harvard Business Review Press      1

[334 rows x 2 columns]

Das 334 editoras que publicaram livros, as maiores são: 
     em primeiro lugar: Penguin Books  
    depois:  
    Vintage	
    Grand Central Publishing	
    Penguin Classics	
    Ballantine Books	
    Bantam	
    Berkley	
    St. Martin's Press	
    Berkley Books	
    William Morrow Paperbacks	
    
    elas publicaram entre 42 e 13 livros

## autor com media mais alta entre mais de 50 classificaçoes <a id='autor'></a>
Identifique o autor com a média mais alta classificação de livros: olhe apenas para livros com pelo menos 50 classificações.
para tanto precisaremos de books e authors e ratings

In [19]:
consulta(books)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [20]:
consulta(authors)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [21]:
consulta(ratings)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

o codigo a seguir é muito complicado então vou explicar:
    na primeira variavel eu armazenei uma junção de tabelas livro e avaliação para calcular a quantidade de avaliações -count(r.rating)- e separar apenas os com mais de 50.
    na segunda variavel eu juntei as tres tabelas livros autores e avaliações juntamente com a consulta da primeira variavel, assim, garanti acesso a quantidade de avaliações, depois filtrei por avalliações >=50 e agrupei por autor calculando a media de avaliações e ordenei em ordem decrescente

In [22]:
livro_avaliacoes = 'SELECT title, COUNT(r.rating) AS numero_avaliacoes FROM books AS b JOIN ratings AS r ON r.book_id = b.book_id GROUP BY b.title'

In [23]:
x=f'select* FROM authors a JOIN books b ON a.author_id=b.author_id JOIN ratings AS r ON r.book_id = b.book_id JOIN  ({livro_avaliacoes}) as livro50 ON b.title = livro50.title where numero_avaliacoes>=50'
consulta(x)

author_id           author  book_id  author_id  \
0           106        Dan Brown       75        106   
1           106        Dan Brown       75        106   
2           106        Dan Brown       75        106   
3           106        Dan Brown       75        106   
4           106        Dan Brown       75        106   
...         ...              ...      ...        ...   
1464        554  Stephenie Meyer      948        554   
1465        554  Stephenie Meyer      948        554   
1466        554  Stephenie Meyer      948        554   
1467        554  Stephenie Meyer      948        554   
1468        554  Stephenie Meyer      948        554   

                                     title  num_pages publication_date  \
0     Angels & Demons (Robert Langdon  #1)        736       2006-04-01   
1     Angels & Demons (Robert Langdon  #1)        736       2006-04-01   
2     Angels & Demons (Robert Langdon  #1)        736       2006-04-01   
3     Angels & Demons (Robert Langdon  #1)        736       2006-04-01   
4     Angels & Demons (Robert Langdon  #1)        736       2006-04-01   
...                                    ...        ...              ...   
1464               Twilight (Twilight  #1)        501       2006-09-06   
1465               Twilight (Twilight  #1)        501       2006-09-06   
1466               Twilight (Twilight  #1)        501       2006-09-06   
1467               Twilight (Twilight  #1)        501       2006-09-06   
1468               Twilight (Twilight  #1)        501       2006-09-06   

      publisher_id  rating_id  book_id       username  rating  \
0              229        389       75     annawarner       5   
1              229        390       75       thomas49       4   
2              229        391       75        julie81       4   
3              229        392       75       robert00       4   
4              229        393       75     jarvispaul       4   
...            ...        ...      ...            ...     ...   
1464           176       6191      948  perrypatricia       3   
1465           176       6192      948     carriehale       3   
1466           176       6193      948       yvasquez       5   
1467           176       6194      948  danielpatrick       4   
1468           176       6195      948    carlsonbrad       5   

                                     title  numero_avaliacoes  
0     Angels & Demons (Robert Langdon  #1)                 84  
1     Angels & Demons (Robert Langdon  #1)                 84  
2     Angels & Demons (Robert Langdon  #1)                 84  
3     Angels & Demons (Robert Langdon  #1)                 84  
4     Angels & Demons (Robert Langdon  #1)                 84  
...                                    ...                ...  
1464               Twilight (Twilight  #1)                160  
1465               Twilight (Twilight  #1)                160  
1466               Twilight (Twilight  #1)                160  
1467               Twilight (Twilight  #1)                160  
1468               Twilight (Twilight  #1)                160  

[1469 rows x 14 columns]

In [24]:
media_avaliacoes = f'SELECT a.author, AVG(r.rating) as avaliacao_media FROM authors a JOIN books b ON a.author_id=b.author_id JOIN ratings AS r ON r.book_id = b.book_id JOIN  ({livro_avaliacoes}) as livro50 ON b.title = livro50.title where numero_avaliacoes>=50 GROUP BY a.author ORDER BY avaliacao_media DESC'
#media_avaliacao_autor=f' select author, avg(rating) from {contando_avaliacoes} as media_avaliacao group by title '
consulta(media_avaliacoes)

author  avaliacao_media
0                          J.K. Rowling/Mary GrandPré         4.287097
1                   Markus Zusak/Cao Xuân Việt Khương         4.264151
2                                      J.R.R. Tolkien         4.246914
3                                   Louisa May Alcott         4.192308
4                                       Arthur Golden         4.107143
5                                        Rick Riordan         4.080645
6                                     William Golding         3.901408
7                                       J.D. Salinger         3.825581
8            Paulo Coelho/Alan R. Clarke/Özdemir İnce         3.789474
9   William Shakespeare/Paul Werstine/Barbara A. M...         3.787879
10                                         Lois Lowry         3.750000
11                                          Dan Brown         3.741259
12            George Orwell/Boris Grabnar/Peter Škerl         3.729730
13                                    Stephenie Meyer         3.662500
14                                     John Steinbeck         3.622951

temos uma avaliação media entre 4.2 e 3.6 alem disso temos 14 autores que tem livros com mais de 50 avaliações. o mais bem avaliado é a JK rowlling em conjunto com a  ilustradora Mary GrandPré depois markus Zusak e depois J.R.R.Tolkien.

[sumario](#sumario)
## media de avaliação dos usuarios com mais de 50 classificaçoes <a id='classificacao'></a>

Encontre o número médio de avaliações entre usuários que classificaram mais do que 50 livros.

In [25]:
consulta(ratings)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [26]:
consulta(reviews)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [27]:
usuarios50mais='select username, count(rating) from ratings group by username having count(rating) >= 50'
consulta(usuarios50mais)

username  count
0     sfitzgerald     55
1  jennifermiller     53
2          xdavis     51
3          paul88     56
4     lesliegibbs     50
5      martinadam     56
6  vanessagardner     50
7       richard89     55
8  shermannatalie     50

contagemusuario = f'SELECT re.username, count(re.text) FROM reviews re JOIN ({usuarios50mais}) AS u50 ON u50.username = re.username WHERE u50.count >= 50 GROUP BY re.username'

contagemusuario= f'select re.username, count(re.text) from reviews re join ({usuarios50mais}) as u50 on u50.username=re.username where u50.count >= 50 group by re.username'


In [28]:
usuarios50mais = 'SELECT username, count(r.rating) FROM ratings r GROUP BY username'

contagemusuario = f'''
    SELECT re.username, count(re.text) 
    FROM reviews re 
    JOIN ({usuarios50mais}) AS u50 ON u50.username = re.username 
    WHERE u50.count >= 50 
    GROUP BY re.username
'''
consulta(contagemusuario)


username  count
0     sfitzgerald     28
1  jennifermiller     25
2          xdavis     18
3          paul88     22
4     lesliegibbs     26
5      martinadam     27
6  vanessagardner     22
7       richard89     26
8  shermannatalie     24

In [29]:
mediausuario = f'''
    SELECT AVG(count) 
    FROM (
        {contagemusuario}
    ) AS subconsulta
'''
consulta(mediausuario)

avg
0  24.222222

esses usuarios mais antigos tem uma media de 24 avaliações/criticas de filmes